src/controller.rs

```diff
@@ -0,0 +1,134 @@
+        self.effect_settings.insert(effect_id, settings_id);
+    }
+
+    pub fn add_connection(&mut self, connection_id: usize, connection: Connection) {
+        self.connections.insert(connection_id, connection);
+    }
+
+    pub fn add_led_strip(&mut self, led_strip_id: usize, led_strip: LedStrip) {
+        self.led_strips.insert(led_strip_id, led_strip);
+    }
+
+    pub fn link_led_strip_to_connection(&mut self, led_strip_id: usize, connection_id: usize) {
+        self.led_strip_connections
+            .insert(led_strip_id, connection_id);
+    }
+
+    pub fn update_led_strips(&mut self) {
+        for (led_strip_id, led_strip) in self.led_strips.iter_mut() {
+            for (effect_id, interval) in &led_strip.effects {
+                let leds = match led_strip.colors.get_mut(interval.0..=interval.1) {
+                    Some(leds) => leds,
+                    None => {
+                        // TODO fix le probleme
+                        log::warn!("Effect {effect_id} has invalid interval ({interval:?}) on ledstrip {led_strip_id} of size {}. Skipping.", led_strip.size);
+                        continue;
+                    }
+                };
+
+                let effect = match self.effects.get_mut(effect_id) {
+                    Some(effect) => effect,
+                    None => {
+                        // TODO fix le probleme
+                        log::warn!("Effect {effect_id} doesn't exist. Skipping.");
+                        continue;
+                    }
+                };
+
+                let setting_id = match self.effect_settings.get(effect_id) {
+                    Some(effect) => effect,
+                    None => {
+                        // TODO fix le probleme
+                        log::warn!("Settings for effect {effect_id} doesn't exist. Skipping.");
+                        continue;
+                    }
+                };
+
+                let setting = self.settings.get(setting_id);
+                match (effect, setting) {
+                    (Effect::Moody(_moody), Some(Settings::Moody(settings))) => {
+                        update_moody(leds, settings);
+                    }
+                    (Effect::Raindrop(raindrop), Some(Settings::Raindrop(settings))) => {
+                        update_raindrop(leds, settings, &mut raindrop.state);
+                    }
+                    (Effect::Lua(lua), Some(Settings::Lua(settings))) => {
+                        if let Err(e) = lua.tick(leds, settings) {
+                            log::error!("Error when executing lua function: {:?}", e);
+                        }
+                    }
+                    _ => panic!("Effect doesn't match settings"),
+                }
+            }
+        }
+    }
+    pub fn send_ledstrip_colors(&mut self) -> anyhow::Result<()> {
+        self.led_strip_connections
+            .retain(|ledstrip_id, connection_id| {
```

Comment:
We use retain to remove tuples that aren't valid anymore.

Reply:
This is O(capacity) instead of O(len) according to the doc. We could use a `to_remove` vector and manually remove them after

Reply:
Aupire c'est chill live